pkg/cloud/services/gc/cleanup.go

```diff
@@ -0,0 +1,143 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/arn"
+	rgapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1beta1"
+	expinfrav1 "sigs.k8s.io/cluster-api-provider-aws/exp/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/annotations"
+)
+
+const (
+	serviceNameTag    = "kubernetes.io/service-name"
+	eksClusterNameTag = "aws:eks:cluster-name"
+)
+
+// Reconcile will perform any setup operations for garbage collection. Default behaviour is to mark
+// a cluster as requiring garbage collection unless it explicitly opts out.
+func (s *Service) Reconcile(ctx context.Context) error {
+	s.scope.Info("reconciling garbage collection")
+
+	val, found := annotations.Get(s.scope.InfraCluster(), annotations.ExternalResourceGCAnnotation)
+	if !found {
+		val = "true"
+	}
+
+	shouldGC, err := strconv.ParseBool(val)
+	if err != nil {
+		return fmt.Errorf("converting value %s of annotation %s to bool: %w", val, annotations.ExternalResourceGCAnnotation, err)
+	}
+
+	if shouldGC {
+		s.scope.V(2).Info("Enabling garbage collection for cluster")
+		controllerutil.AddFinalizer(s.scope.InfraCluster(), expinfrav1.ExternalResourceGCFinalizer)
+
+		if patchErr := s.scope.PatchObject(); patchErr != nil {
+			return fmt.Errorf("patching infra cluster after adding gc finalizer: %w", patchErr)
+		}
+	}
+
+	return nil
+}
+
+// ReconcileDelete performs any operations that relate to the reconciliation of a cluster delete as it relates to
+// the external resources created by the workload cluster. For example, it will delete the ELB/NLBs that where created
+// as a result of Services of type load balancer.
+func (s *Service) ReconcileDelete(ctx context.Context) error {
+	s.scope.Info("reconciling deletion for garbage collection")
+
+	if err := s.deleteResources(ctx); err != nil {
+		return fmt.Errorf("deleting workload services of type load balancer: %w", err)
+	}
+
+	s.scope.V(2).Info("Removing garbage collection finalizer cluster")
+	controllerutil.RemoveFinalizer(s.scope.InfraCluster(), expinfrav1.ExternalResourceGCFinalizer)
+
+	if patchErr := s.scope.PatchObject(); patchErr != nil {
+		return fmt.Errorf("patching infra cluster after removing gc finalizer: %w", patchErr)
+	}
+
+	return nil
+}
+
+func (s *Service) deleteResources(ctx context.Context) error {
+	s.scope.Info("deleting aws resources created by tenant cluster")
+
+	serviceTag := infrav1.ClusterAWSCloudProviderTagKey(s.scope.KubernetesClusterName())
+	awsInput := rgapi.GetResourcesInput{
+		TagFilters: []*rgapi.TagFilter{
```

Comment:
```suggestion
		ResourceTypeFilters: nil,
		TagFilters: []*rgapi.TagFilter{
```

I'm suggesting this because making it explicit helps me understand that this selects _all_ resources.

I also think that (in a subsequent PR) it would be helpful to factor out a common implementation of this useful, high-level concept of "find resources that are owned by the cluster." We have a similar implementation in one other place: https://github.com/kubernetes-sigs/cluster-api-provider-aws//blob/4d3a8c63fab99a3aa9c029394bce08cf343b9855/pkg/cloud/services/elb/loadbalancer.go#L593-L601

Reply:
Changed to add `ResourceTypeFilters: nil,`

I agree lets refactor the resource tagging into a common function in a later PR.